### PR TITLE
Extract ClusterManager from DomainIndex and make it dynamic

### PR DESCRIPTION
**Pull request by PunkPun**
I extract an abstract class `ClusterManager` from `DomainIndex` and make it fully dynamic. Old `DomainIndex` didn't properly update domains and that's the reason the map Poland Raid was so laggy for AI

<details>
<summary>demonstration of the `ClusterManager`</summary>

https://user-images.githubusercontent.com/37534529/152613964-c51e41e6-65df-4726-9567-5f02cb135f67.mp4

</details>

How it works:
- `ClusterManager` has an abstract bool `ClusterCondition(CPos)` that is used to define a cell as empty or full. 
- When you build the clusters with `BuildDomain()` it creates cluster classes `ClusterContents`. `ClusterManager` has a `CellLayer<ClusterContents>` that way checking whether 2 cells are in the same cluster remains very performance inexpensive. `ClusterContents` saves all cells that belong in it in `List<CPos>`, so when you need to get all cells in a domain you can easily do it. It's especially useful when updating
- `ClusterManager` additionally contains `List<ClusterContents>, so getting all clusters is also very easy. I define an empty `ClusterContents`, it contains all cells that don't satisfy `ClusterCondition`. `ClusterContents` has a bool `IsEmpty`. With this you can check wether a cell is in a cluster or empty
- When updating a cell I check `ClusterCondition` to see what I need to do. If cell was added, I look for nearby clusters. If none were found I create a new one. I found one I join it. If I found multiple I join the first one and assimilate the rest into the first one. When a cell is removed I check for splitting and the recalculate the cluster and remove more the cell to the empty cluster

If you wanted to cluster harvestable resources, you'd create a class that inherits `ClusterManager` and for every resource type (eg. gold and gems) you'd create an instance of it. Then if you wanted to get the value of a cluster when hovering with a mouse, you'd go through every `ClusterManager` checking their CellLayer for a `IsEmpty == false` cluster. If you are actually hovering on a resource you will find `ClusterContents` that contains an int Cash value

The reason why this is a draft is because II haven't completely finished the code. I have a few questions

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+						.Where(d => visited.Contains(d) && !visited[d])
+						.Do(d => clusterQueue.Enqueue(d));
+				}
+			}
+
+			return cluster.ClusterId;
+		}
+	}
+
+	class ClusterContents
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs:278]: What would be the best way to make this dynamic for openra engine? I'd like you to be able to give it custom code and values, like for the resource example it should hold the `Cash` value

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+		public ClusterContents(bool empty)
+		{
+			ClusterId = 0;
+			IsEmpty = true;
+		}
+	}
+
+	// For TS bridges and tunnels
+	class ClusterTransientConnection
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs:304]: I still haven't tested if they are updated dynamically, I need to create a good testcase. 
Another note, since these will likely be only used for TS bridges and tunnels should I extract this code to DomainIndex?

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/WorldTooltipLogic.cs [hunk 3]
```diff
@@ -80,6 +81,8 @@ public WorldTooltipLogic(Widget widget, World world, TooltipContainerWidget tool
 						}
 				}
 
+				labelText += " " + world.WorldActor.Trait<DomainIndex>().GetDomain(viewport.Cell);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/WorldTooltipLogic.cs:84]: You can hover over actors to see domain index. This is used to test if clusters are being updated accordingly, the map Minor Pressure can be used as testcase

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+			foreach (var cell in removedCells)
+			{
+				// We don't need to test for splitting if it's the last cell in a domain
+				if (clusterCellLayer[cell].Cells.Count > 1)
+				{
+					// Select all neighbors inside the map boundaries
+					var neighbors = CVec.Directions.Select(d => d + cell)
+						.Where(c => clusterCellLayer[cell].Cells.Contains(c) && !removedCells.Contains(c)).ToList();
+
+					// We don't need to test for splitting if it's an edge of a domain
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs:97]: This needs to be a smarter check, we don't want to be recalculating the whole cluster when it is clearly not needed

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+			return cluster.ClusterId;
+		}
+	}
+
+	class ClusterContents
+	{
+		public readonly bool IsEmpty;
+
+		// For debugging only
+		static ushort domainIndex = 1;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs:283]: `domainIndex` is only used for debugging, when we get around to merging we may want to remove it

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+		public readonly List<ClusterTransientConnection> TransientConnections = new List<ClusterTransientConnection>();
+
+		public ClusterContents()
+		{
+			ClusterId = domainIndex;
+			domainIndex += 1;
+		}
+
+		public ClusterContents(bool empty)
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs]: I'm not sure what the best way to create the empty ClusterContents is

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+				}
+
+				UpdateCell(cell, emptyCluster);
+			}
+		}
+
+		public void AddFixedConnection(IEnumerable<CPos> connectedCells)
+		{
+			var cells = connectedCells.ToList();
+			var connection = new ClusterTransientConnection(true, cells);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/ClusterManager.cs:109]: TS bridge connections can be turned on and off. Temporarily they are all set to to true

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs]: Made it private set public get for now

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs (hunk 4 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:97]: IIrc that was the idea behind additionally dividing the map into chunks.

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:278]: You could make it hold a reference to a type (i.e. `ClusterContents<T>`), which should cover the use case of holding cash?

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:278]: Yes, and in a similar theme to `ClusterCondition` you could provide a `protected abstract T ClusterState(IEnumerable<CPos> cells);` so the user can specify how to determine the value. It could be called whenever you update the cells in a cluster.

Or you could go more granular and have methods for addition, removal, etc.

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs (hunk 4 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:97]: Destroying a bridge on the Hypothermia map in Red Alert freezes the game for ~3 seconds whilst it does this recalculation. The current domain index avoids the performance issue by being incorrect and not updating at all.

Even if we divided the area up into grids, we still suffer a large costs because we are trying to maintain an up-to-date list of cells within each domain. When you need to split a large domain into two, there's no way to get around the fact you're going to have to update some large lists.

The hierarchical path finder divides the map up into grids, and tracks connectivity at that level. If a split occurs, it only has to rebuild the connections on that grid. It gets away with that because it is *not* trying to keep a specific list of cells in each domain index. That means if you want to know if cells belong to the same domain index, you still need to invoke a path search to see if they are connected! The "hierarchical" part means this is reasonably cheap because you can consult the much smaller abstract graph to determine connected regions, rather than having to consult the full sized map.

The hierarchical path finder would allow us to replace queries against the domain index that are related to pathing, e.g. when the naval AI is trying to find the nearest enemy to attack that is reachable over water. For other considerations like this PR was working towards such as tracking clusters of ore and gems, we'd need to separate out the "hierarchical" logic from the "pathfinding" logic in order to reuse it.

Coming back to focus on this PR on its own: I'm wondering if making the domain index correct by having it handle more updates is worth the performance cost that incurs in the short term. Given we've gotten away with it for years maybe the lesser evil is to keep it as is, much as that pains me to suggest as this is a nice PR conceptually.

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:97]: The recalculation is not yet optimised well, it has a few glaring problems that I'll address hopefully soon. When testing I haven't experienced any lag-spikes but I guess I now have a good testcase.

**Comment by PunkPun** [comment]
related #10730

File: OpenRA.Mods.Common/Traits/World/ClusterManager.cs (hunk 5 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/World/ClusterManager.cs:283]: DomainIndex was removed in #19591

**Comment by RoosterDragon** [comment]
@PunkPun with #19591 merged I would suggest this PR is now obsolete.

**Comment by PunkPun** [comment]
> @PunkPun with #19591 merged I would suggest this PR is now obsolete.

It's not obsolete, though the name definitely is. We still need a generic domain index for other uses like calculating cash in orefields for map editor, or perhaps for AI to understand what is an orefield and how to expand to it etc. I'm not going to be working on this any time soon so maybe it's better to just open an issue and link this PR to it (also close this PR)

**Comment by PunkPun** [comment]
Closing and adding to `Future` milestone
